In [1]:
import pandas as pd
import sqlite3

# 8.1 Reading data from SQL databases

So far we've only talked about reading data from CSV files. That's a pretty common way to store data, but there are many others! Pandas can read from HTML, JSON, SQL, Excel (!!!), HDF5, Stata, and a few other things. In this chapter we'll talk about reading data from SQL databases.

You can read data from a SQL database using the `pd.read_sql` function. `read_sql` will automatically convert SQL column names to DataFrame column names.

`read_sql` takes 2 arguments: a `SELECT` statement, and a database connection object. This is great because it means you can read from *any* kind of SQL database -- it doesn't matter if it's MySQL, SQLite, PostgreSQL, or something else.

This example reads from a SQLite database, but any other database would work the same way.

In [2]:
con = sqlite3.connect("./data/weather_2012.sqlite")
df = pd.read_sql("SELECT * from weather_2012 LIMIT 30", con)
df

id            date_time  temp
0    1  2012-01-01 00:00:00  -1.8
1    2  2012-01-01 01:00:00  -1.8
2    3  2012-01-01 02:00:00  -1.8
3    4  2012-01-01 03:00:00  -1.5
4    5  2012-01-01 04:00:00  -1.5
5    6  2012-01-01 05:00:00  -1.4
6    7  2012-01-01 06:00:00  -1.5
7    8  2012-01-01 07:00:00  -1.4
8    9  2012-01-01 08:00:00  -1.4
9   10  2012-01-01 09:00:00  -1.3
10  11  2012-01-01 10:00:00  -1.0
11  12  2012-01-01 11:00:00  -0.5
12  13  2012-01-01 12:00:00  -0.2
13  14  2012-01-01 13:00:00   0.2
14  15  2012-01-01 14:00:00   0.8
15  16  2012-01-01 15:00:00   1.8
16  17  2012-01-01 16:00:00   2.6
17  18  2012-01-01 17:00:00   3.0
18  19  2012-01-01 18:00:00   3.8
19  20  2012-01-01 19:00:00   3.1
20  21  2012-01-01 20:00:00   3.2
21  22  2012-01-01 21:00:00   4.0
22  23  2012-01-01 22:00:00   4.4
23  24  2012-01-01 23:00:00   5.3
24  25  2012-01-02 00:00:00   5.2
25  26  2012-01-02 01:00:00   4.6
26  27  2012-01-02 02:00:00   3.9
27  28  2012-01-02 03:00:00   3.7
28  29  2012-01-02 04:00:00   2.9
29  30  2012-01-02 05:00:00   2.6

`read_sql` doesn't automatically set the primary key (`id`) to be the index of the dataframe. You can make it do that by adding an `index_col` argument to `read_sql`. 

If you've used `read_csv` a lot, you may have seen that it has an `index_col` argument as well. This one behaves the same way.

In [3]:
df = pd.read_sql("SELECT * from weather_2012 LIMIT 30", con, index_col='id')
df

date_time  temp
id                           
1   2012-01-01 00:00:00  -1.8
2   2012-01-01 01:00:00  -1.8
3   2012-01-01 02:00:00  -1.8
4   2012-01-01 03:00:00  -1.5
5   2012-01-01 04:00:00  -1.5
6   2012-01-01 05:00:00  -1.4
7   2012-01-01 06:00:00  -1.5
8   2012-01-01 07:00:00  -1.4
9   2012-01-01 08:00:00  -1.4
10  2012-01-01 09:00:00  -1.3
11  2012-01-01 10:00:00  -1.0
12  2012-01-01 11:00:00  -0.5
13  2012-01-01 12:00:00  -0.2
14  2012-01-01 13:00:00   0.2
15  2012-01-01 14:00:00   0.8
16  2012-01-01 15:00:00   1.8
17  2012-01-01 16:00:00   2.6
18  2012-01-01 17:00:00   3.0
19  2012-01-01 18:00:00   3.8
20  2012-01-01 19:00:00   3.1
21  2012-01-01 20:00:00   3.2
22  2012-01-01 21:00:00   4.0
23  2012-01-01 22:00:00   4.4
24  2012-01-01 23:00:00   5.3
25  2012-01-02 00:00:00   5.2
26  2012-01-02 01:00:00   4.6
27  2012-01-02 02:00:00   3.9
28  2012-01-02 03:00:00   3.7
29  2012-01-02 04:00:00   2.9
30  2012-01-02 05:00:00   2.6

If you want your dataframe to be indexed by more than one column, you can give a list of columns to `index_col`:

In [4]:
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con, 
                 index_col=['id', 'date_time'])
df

temp
id date_time                
1  2012-01-01 00:00:00  -1.8
2  2012-01-01 01:00:00  -1.8
3  2012-01-01 02:00:00  -1.8

# 8.2 Writing to a SQLite database

Pandas has a `DataFrame.to_sql` function which creates a database table from a dataframe. Let's use it to move our 2012 weather data into SQL.

In [5]:
weather_df = pd.read_csv('./data/weather_2012.csv')
con = sqlite3.connect("./data/test_db.sqlite")
con.execute("DROP TABLE IF EXISTS weather_2012")
weather_df.to_sql("weather_2012", con)

/Users/bikramkundu/opt/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py:2779: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  sql.to_sql(


We can now read from the `weather_2012` table in  `test_db.sqlite`, and we see that we get the same data back:

In [7]:
con = sqlite3.connect("./data/test_db.sqlite")
df = pd.read_sql("SELECT * from weather_2012 LIMIT 10", con)
df

index            Date/Time  Temp (C)  Dew Point Temp (C)  Rel Hum (%)  \
0      0  2012-01-01 00:00:00      -1.8                -3.9           86   
1      1  2012-01-01 01:00:00      -1.8                -3.7           87   
2      2  2012-01-01 02:00:00      -1.8                -3.4           89   
3      3  2012-01-01 03:00:00      -1.5                -3.2           88   
4      4  2012-01-01 04:00:00      -1.5                -3.3           88   
5      5  2012-01-01 05:00:00      -1.4                -3.3           87   
6      6  2012-01-01 06:00:00      -1.5                -3.1           89   
7      7  2012-01-01 07:00:00      -1.4                -3.6           85   
8      8  2012-01-01 08:00:00      -1.4                -3.6           85   
9      9  2012-01-01 09:00:00      -1.3                -3.1           88   

   Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)               Weather  
0                4              8.0           101.24                   Fog  
1                4              8.0           101.24                   Fog  
2                7              4.0           101.26  Freezing Drizzle,Fog  
3                6              4.0           101.27  Freezing Drizzle,Fog  
4                7              4.8           101.23                   Fog  
5                9              6.4           101.27                   Fog  
6                7              6.4           101.29                   Fog  
7                7              8.0           101.26                   Fog  
8                9              8.0           101.23                   Fog  
9               15              4.0           101.20                   Fog

The nice thing about having your data in a database is that you can do arbitrary SQL queries. This is cool especially if you're more familiar with SQL. Here's an example of sorting by the Weather column:

In [ ]:
con = sqlite3.connect("./data/test_db.sqlite")
df = pd.read_sql("SELECT * from weather_2012 ORDER BY Weather LIMIT 3", con)
df

If you have a PostgreSQL database or MySQL database, reading from it works exactly the same way as reading from a SQLite database. You create a connection using `psycopg2.connect()` or `MySQLdb.connect()`, and then use

`pd.read_sql("SELECT whatever from your_table", con)`

# 8.3 Connecting to other kinds of database - MySQL

In [ ]:
!pip install pymysql

### Writing to MySQL Table

*Note: For these to work, you will need a working MySQL / PostgreSQL database, with the correct localhost, database name, etc.*

In [13]:
from sqlalchemy import create_engine
import pymysql

tableName   = "diamonds"
df   = pd.read_csv('./data/diamonds.csv')
sqlEngine   = create_engine('mysql+pymysql://root:root@127.0.0.1/pandas_test', pool_recycle=3600)
dbConnection= sqlEngine.connect()
 
try:
    df = df.to_sql(tableName, dbConnection, if_exists='fail');
except ValueError as vx:
    print(vx)
except Exception as ex:   
    print(ex)

else:
    print("Table %s created successfully."%tableName);   

finally:

    dbConnection.close()

Table diamonds created successfully.


### And Reading From MySQL Table

In [17]:
from sqlalchemy import create_engine
import pymysql

sqlEngine    = create_engine('mysql+pymysql://root:root@127.0.0.1/pandas_test', pool_recycle=3600)
dbConnection = sqlEngine.connect()
df           = pd.read_sql("select * from diamonds", dbConnection);

pd.set_option('display.expand_frame_repr', False)
df = df.set_index('index')
display(df)
dbConnection.close()

carat        cut clarity  price
index                                 
0       0.23      Ideal     SI2    326
1       0.21    Premium     SI1    326
2       0.23       Good     VS1    327
3       0.29    Premium     VS2    334
4       0.31       Good     SI2    335
...      ...        ...     ...    ...
53935   0.72      Ideal     SI1   2757
53936   0.72       Good     SI1   2757
53937   0.70  Very Good     SI1   2757
53938   0.86    Premium     SI2   2757
53939   0.75      Ideal     SI2   2757

[53940 rows x 4 columns]

### Credit:
Some of the content examples are copied from
- https://github.com/jvns/pandas-cookbook/tree/master/cookbook
- https://pythontic.com/pandas/serialization/mysql

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  